### ISSUE TITLE
no-unneeded-ternary does not detect a ? a : b

### ISSUE THREAD

**Issue by binury** [eslint/eslint#14190]
**What rule do you want to change?**
no-unneeded-ternary

**Does this change cause the rule to produce more or fewer warnings?**
more

**How will the change be implemented? (New option, new default behavior, etc.)?**
The existing checks on ConditionalExpression-nodes should be amended to report a problem when the test and consequent of a ternary expression are the same (`a ? a : b`)

**Please provide some example code that this change will affect:**

```js
// Bad
bar ? bar : 1

// Good
bar || 1

```

**What does the rule currently do for this code?**
**[The documentation for `no-unneeded-ternary`](https://eslint.org/docs/rules/no-unneeded-ternary) describes this problem but doesn't appear to address it,** almost as if it was intended as a TODO
> Another common mistake is using a single variable as both the conditional test and the consequent. In such cases, the logical OR can be used to provide the same functionality

**What will the rule do after it's changed?**
no-unneeded-ternary should detect short-circuits masquerading as ternaries (the conditional and consequent are identical)

**Are you willing to submit a pull request to implement this change?**
👍

**Comment by binury** [comment]
Actually it seems like the rule has some code in place to test this…
https://github.com/eslint/eslint/blob/1050ee78a95da9484ff333dc1c74dac64c05da6f/lib/rules/no-unneeded-ternary.js#L107-L117

I was under the impression it did not. The documentation (notably dated) describes this use-case but it doesn't have any concrete examples which make use of the rule.

Will check the rule's tests and also look at why this rule isn't applying on my project; hopefully not a local problem 😅

**Comment by binury** [comment]
Looks like it's actually disabled by default but can be enabled with an option!

**Comment by mdjermanovic** [comment]
Hi @binury, thanks for the issue!

This check is not enabled by default. You can set [`defaultAssignment: false`](https://eslint.org/docs/rules/no-unneeded-ternary#defaultassignment) option to enable it:

```js
/* eslint no-unneeded-ternary: ["error", { defaultAssignment: false }] */

bar ? bar : 1 // error
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVubmVlZGVkLXRlcm5hcnk6IFtcImVycm9yXCIsIHsgZGVmYXVsdEFzc2lnbm1lbnQ6IGZhbHNlIH1dICovXG5cbmJhciA/IGJhciA6IDEgLy8gZXJyb3IiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)
